plutus-core/executables/debugger/Debugger.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Logic for the debugging stage.
+module Debugger where
+
+import Common
+import Types
+
+import Brick.Focus qualified as B
+import Brick.Types qualified as B
+import Brick.Widgets.Border qualified as BB
+import Brick.Widgets.Center qualified as BC
+import Brick.Widgets.Core qualified as B
+import Brick.Widgets.Edit qualified as BE
+import Control.Monad.State
+import Data.Bifunctor
+import Data.Maybe
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Graphics.Vty qualified as Vty
+import Lens.Micro
+
+drawDebugger ::
+    DebuggerState ->
+    [B.Widget ResourceName]
+drawDebugger st = [header "Plutus Core Debugger" $ BC.center ui]
+  where
+    focusRing = st ^. dsFocusRing
+    (curRow, curCol) = BE.getCursorPosition (st ^. dsUplcEditor)
+    cursorPos = "Ln " <> show (curRow + 1) <> ", Col " <> show (curCol + 1)
+    uplcEditor =
+        BB.borderWithLabel (B.txt "UPLC program") $
+            B.vBox
+                [ B.withFocusRing
+                    focusRing
+                    (BE.renderEditor drawTexts)
+                    (st ^. dsUplcEditor)
+                , B.padTop (B.Pad 1) . BC.hCenter $ B.str cursorPos
+                ]
+    sourceEditor =
+        BB.borderWithLabel (B.txt "Source program") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsSourceEditor)
+    returnValueEditor =
+        BB.borderWithLabel (B.txt "UPLC value being returned") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsReturnValueEditor)
+    cekStateEditor =
+        BB.borderWithLabel (B.txt "CEK machine state") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsCekStateEditor)
+    ui =
+        B.vBox
+            [ BC.center uplcEditor B.<+> BC.center sourceEditor
+            , B.vLimit (st ^. dsVLimitBottomEditors) $
+                BC.center returnValueEditor B.<+> BC.center cekStateEditor
+            , footer
+            ]
+    drawTexts :: [Text] -> B.Widget ResourceName
```

Comment:
A bit vague and in fact it's quite specific: this is drawing a _document_ (i.e. a list of lines) with support for highlighting. Could even be promoted to a top-level function taking the highlighting options I think, and it would be clearer. I could have done with a comment explaining the general approach here (treat the document as a list of widgets, one for each row, so we can individually highlight the selected row).